packages/react-form-renderer/src/form-renderer/form-renderer.js

```diff
@@ -26,9 +26,21 @@ const FormRenderer = ({
   ...props
 }) => {
   const [fileInputs, setFileInputs] = useState([]);
+  const registeredFields = useRef({});
   const focusDecorator = useRef(createFocusDecorator());
   let schemaError;
 
+  const setRegisteredFields = (fn => registeredFields.current = fn({...registeredFields.current}));
+  const internalRegisterField = (name) => {
+    setRegisteredFields(prev => prev[name] ? ({...prev, [name]: prev[name] + 1}) : ({...prev, [name]: 1}));
+  };
+
+  const internalUnRegisterField = (name) => {
+    setRegisteredFields(({[name]: currentField, ...prev}) => currentField && currentField > 1 ? ({[name]: currentField - 1, ...prev}) : prev);
+  };
+
+  const internalGetRegisteredFields = () => Object.entries(registeredFields.current).reduce((acc, [name, value]) => value > 0 ? [...acc, name] : acc, []);
+
```

Comment:
What about to do what I proposed (this seems unnecessary complex to me):

```jsx
  const registeredFields = useRef([]);
```

register: 
```jsx
const internalRegisterField = (name) => registeredFields.current.push(name)
```

unregister: 

```jsx
function removeElement(array, elem) {
    var index = array.indexOf(elem);
    if (index > -1) {
        array.splice(index, 1);
    }
}

const internalUnRegisterField = (name) => removeElement(registeredFields.current, name)
```

internalGetRegisteredFields

```jsx
const internalGetRegisteredFields = () => new Set(registeredFields.current)
```

Reply:
There is a couple of potential issues. First, I don't like the idea of having the same item multiple times inside the array. Using an object is slightly more resource-efficient than looking for indexes.

Another problem might be the usage of `push` which can create an out-of-sync structure when we re-create the array (splice in removeElement). Theoretically, we can push into an array that no longer exists.

Also this:
```jsx
const internalGetRegisteredFields = () => new Set(registeredFields.current)
```

Would return an instance of a set instead of an array. We would need either `Array.from` or `[...x]` destructor to actually get an array to preserve current API. So we need some value transformation anyway.